build-logic/packaging/src/main/kotlin/gradlebuild/instrumentation/transforms/InstrumentationMetadataTransform.kt

```diff
@@ -129,4 +115,14 @@ abstract class InstrumentationMetadataTransform : TransformAction<TransformParam
             REMOVED -> null
         }
     }
+
+    private
+    fun File?.copyTo(other: File, defaultValue: String) {
```

Comment:
🤔 This is weird and counter-intuitive. Can we instead do this without `?` and call it with:

```kotlin
newInstrumentedClassesFile?.copyTo(instrumentedClassesFile, defaultValue = "")
```

I think that would be a lot better, because it clearly states at the call that we might not do anything if `newInstrumentedClassesFile` is `null`.

Reply:
Actually when there is `null` we write empty/default value to the output. Current "contract" of changes is a bit hidden.
When:
- `newInstrumentedClassesFile == null || !newInstrumentedClassesFile.exists()` -> that means nothing is upgraded, so we need to write empty value in case something was upgraded before
- `newInstrumentedClassesFile == oldUpgradedPropertiesFile` -> means nothing changed to upgrades file
- `newInstrumentedClassesFile != oldUpgradedPropertiesFile` -> means upgrades file changed

We could represent better this states via some enums or types. E.g. from `findChanges` we would return not just a file, but some wrapper around a file that would make this logic more obvious